### PR TITLE
Implement of trait 'Dataset' and struct 'MNIST'

### PR DESCRIPTION
**Pull request by litcoderr**
[Done]
- Defined trait 'Dataset' methods
  - len()
  - get_item()
- Made built-in MNIST dataset
    - Download file from link
    - Decompress file
    - Implemented `len()` and `get_item()` from trait `Dataset`
    - Implemented trait `Debug` for CLI visualization
image

[Todo]
- Make `Dataloader` for multi-processor batching sequence
- Specify trait `DatasetItem` when `Tensor` is usable

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,5 @@ Cargo.lock
 
 # These are backup files generated by rustfmt
 **/*.rs.bk
+
+raw
```
Comment (utilForever) [inline .gitignore:13]: What is this? Can you add comment for the meaning of this line?

File: src/utils/data.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
+use flate2::read::GzDecoder;
+use std::error;
+use std::io::{Cursor, Read};
+use std::path::PathBuf;
+
+pub mod dataset;
+
+/// Downloads file from url to designated root path.
+/// Downloading process will only be proceeded when file path does not exist
```
Comment (utilForever) [inline src/utils/data.rs]:
```suggestion
/// Download process will be proceeded when the file path does not exist.
```

File: src/utils/data.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+///
+/// # Arguments
+/// * `root` - Root directory of dataset
+/// * `file_name` - Name of file
+/// * `url` - Url where original data is stored
```
Comment (utilForever) [inline src/utils/data.rs]:
```suggestion
/// * `root` - The root directory of dataset
/// * `file_name` - The file name to download
/// * `url` - URL where the original data is stored
```

File: src/utils/data.rs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+pub fn download_from_url(
+    root: &mut PathBuf,
+    file_name: &str,
+    url: &str,
+    decompress: bool,
```
Comment (utilForever) [inline src/utils/data.rs]: 1. Where is the comment for parameter `decompress`?
2. I suggest to rename this parameter.

```suggestion
    is_decompress: bool,
```

File: src/utils/data.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            std::io::copy(&mut content, &mut file)?;
+        } else {
+            // write bytes to file
+            let mut content = Cursor::new(resp.bytes()?);
+            std::io::copy(&mut content, &mut file)?;
+        }
+    }
+    // convert back to original data root path
+    root.pop();
+    Ok(())
```
Comment (utilForever) [inline src/utils/data.rs]: I think that line-by-line comments are unnecessary.
Minimize comments by improving readability such as variable renaming.

File: src/utils/data.rs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+/// Convert byte array to u32
+///
+/// # Arguments
+/// * `byt` : byte array
+/// * `offset` : offset of byte array to convert
+/// * `big_endian` : Boolean for if data is stored as big_endian
```
Comment (utilForever) [inline src/utils/data.rs]:
```suggestion
/// * `byte_arr` : The original byte array
/// * `offset` : The offset of the byte array to convert
/// * `is_big_endian` : A flag indicates the data is stored as big endian
```

File: src/utils/data.rs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+fn byte_arr_to_u32(byt: &Vec<u8>, offset: u32, big_endian: bool) -> u32 {
```
Comment (utilForever) [inline src/utils/data.rs]:
```suggestion
fn convert_byte_arr_to_u32(byte_arr: &Vec<u8>, offset: u32, is_big_endian: bool) -> u32 {
```

File: src/utils/data/dataset.rs [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    use std::path::PathBuf;
+    use std::{error, fmt, fs};
+
+    /// On Memory Data struct
+    ///
+    /// # Parameters
+    /// * `root` - Root directory
+    /// * `file_name` - name of downloaded file
+    /// * `source` - file online source
+    /// * `raw` - raw data on memory
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
    /// * `raw` - The raw data on memory
```

File: src/utils/data/dataset.rs [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+    struct OnMemData {
+        raw: Vec<u8>,
+    }
+    impl OnMemData {
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
    }
    
    impl OnMemData {
```

File: src/utils/data/dataset.rs [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+        /// Constructor
+        ///
+        /// # Arguments
+        /// * `root` - Root directory
+        /// * `file_name` - name of downloaded file
+        /// * `source` - file online source
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
        /// * `root` - The root directory
        /// * `file_name` - The downloaded file name
        /// * `source` - The source URL to download
```

File: src/utils/data/dataset.rs [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+        fn new(
+            mut root: PathBuf,
+            file_name: &'static str,
+            source: &'static str,
+        ) -> Result<OnMemData, Box<dyn error::Error>> {
+            // Download if file does not exist
+            download_from_url(&mut root, file_name, source, true)?;
+
+            // load file to memory
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
            // Load file to memory
```

File: src/utils/data/dataset.rs [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+        }
+    }
+
+    /// Mnist Item
+    ///
+    /// # Parameters
+    /// * `image` - image vector
+    /// * `label` - label (ranging from 0 to 9)
+    /// * `row` - row size
+    /// * `col` - col size
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
    /// * `image` - The image vector
    /// * `label` - The label (ranging from 0 to 9)
    /// * `row` - The row size
```

File: src/utils/data/dataset.rs [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+    pub struct MnistItem {
+        image: Vec<u8>,
+        label: u8,
+        row: u32,
+    }
+    impl DatasetItem for MnistItem {}
+    impl fmt::Debug for MnistItem {
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
    }
    
    impl DatasetItem for MnistItem {}
    
    impl fmt::Debug for MnistItem {
```

File: src/utils/data/dataset.rs [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+            let label_str: String = format!("label: {}\n", self.label);
+            let mut image_str: String = format!("image:\n");
+
+            for (i, pixel) in self.image.iter().enumerate() {
+                if *pixel == 0 {
+                    image_str.push_str(" ");
+                } else {
+                    image_str.push_str("*");
+                }
+                if (i + 1) as u32 % self.row == 0 {
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
                }
                
                if (i + 1) as u32 % self.row == 0 {
```

File: src/utils/data/dataset.rs [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+                    image_str.push_str("\n");
+                }
+            }
+            write!(f, "{}{}", label_str, image_str)
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
            }
            
            write!(f, "{}{}", label_str, image_str)
```

File: src/utils/data/dataset.rs [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+        }
+    }
+
+    /// MNIST dataset
+    ///
+    /// # Parameters
+    /// * `root` - Root directory of dataset
```
Comment (utilForever) [inline src/utils/data/dataset.rs]: Incorrect comments for parameters. Please rewrite it.

File: src/utils/data/dataset.rs [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+    pub struct MNIST {
+        test: bool,
+        train_img: Option<OnMemData>,
+        train_label: Option<OnMemData>,
+        test_img: Option<OnMemData>,
+        test_label: Option<OnMemData>,
+    }
+    impl MNIST {
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
    }
    
    impl MNIST {
```

File: src/utils/data/dataset.rs [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+        /// Constructor of MNIST dataset
+        ///
+        /// # Arguments
+        /// * `root` - Root directory of dataset
+        /// * `train` - If True, downloads training set, otherwise downloads test set
+        fn new(root: PathBuf, test: bool) -> Result<MNIST, Box<dyn error::Error>> {
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
        /// * `root` - The root directory of dataset
        /// * `is_test` - true: downloads test set, false: downloads train set
        fn new(root: PathBuf, is_test: bool) -> Result<MNIST, Box<dyn error::Error>> {
```

File: src/utils/data/dataset.rs [hunk 19]
```diff
@@ -0,0 +1,277 @@
...
+                    root.clone(),
+                    "mnist_test_img",
+                    "http://yann.lecun.com/exdb/mnist/t10k-images-idx3-ubyte.gz",
+                )?);
+                let test_label = Some(OnMemData::new(
+                    root.clone(),
+                    "mnist_test_label",
+                    "http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz",
+                )?);
+                Ok(MNIST {
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
                )?);
                
                Ok(MNIST {
```

File: src/utils/data/dataset.rs [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+                    root.clone(),
+                    "mnist_train_img",
+                    "http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz",
+                )?);
+                let train_label = Some(OnMemData::new(
+                    root.clone(),
+                    "mnist_train_label",
+                    "http://yann.lecun.com/exdb/mnist/train-labels-idx1-ubyte.gz",
+                )?);
+                Ok(MNIST {
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
                )?);
                
                Ok(MNIST {
```

File: src/utils/data/dataset.rs [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+                    train_img,
+                    train_label,
+                    test_img: None,
+                    test_label: None,
+                })
+            }
+        }
+
+        /// Returns dimension of each image.
+        /// Returns [N_ROW, N_COL] vector
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
        /// Returns [N_ROW, N_COL] vector of each image.
```

File: src/utils/data/dataset.rs [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+        fn get_dim(&self) -> Vec<u32> {
+            let mut image_data: Option<&OnMemData> = None;
+
+            if self.test {
+                if let Some(image_) = &self.test_img {
+                    image_data = Some(image_);
+                }
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
                if let Some(image) = &self.test_img {
                    image_data = Some(image);
                }
```

File: src/utils/data/dataset.rs [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+            } else {
+                if let Some(image_) = &self.train_img {
+                    image_data = Some(image_);
+                }
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
                if let Some(image) = &self.train_img {
                    image_data = Some(image);
                }
```

File: src/utils/data/dataset.rs [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+            }
+
+            let mut n_rows: u32 = 0;
+            let mut n_cols: u32 = 0;
+            if let Some(image_) = image_data {
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
            let mut n_cols: u32 = 0;
            
            if let Some(image) = image_data {
```

File: src/utils/data/dataset.rs [hunk 25]
```diff
@@ -0,0 +1,277 @@
...
+                n_rows = byte_arr_to_u32(&image_.raw, 8, true);
+                n_cols = byte_arr_to_u32(&image_.raw, 12, true);
+            }
+            vec![n_rows, n_cols]
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
            }
            
            vec![n_rows, n_cols]
```

File: src/utils/data/dataset.rs [hunk 26]
```diff
@@ -0,0 +1,277 @@
...
+                if let Some(image_) = &self.train_img {
+                    image_data = Some(image_);
+                }
+            }
+
+            let mut n_rows: u32 = 0;
+            let mut n_cols: u32 = 0;
+            if let Some(image_) = image_data {
+                n_rows = byte_arr_to_u32(&image_.raw, 8, true);
+                n_cols = byte_arr_to_u32(&image_.raw, 12, true);
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
                n_rows = byte_arr_to_u32(&image.raw, 8, true);
                n_cols = byte_arr_to_u32(&image.raw, 12, true);
```

File: src/utils/data/dataset.rs [hunk 27]
```diff
@@ -0,0 +1,277 @@
...
+            }
+            vec![n_rows, n_cols]
+        }
+    }
+    impl Dataset<MnistItem> for MNIST {
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
    }
    
    impl Dataset<MnistItem> for MNIST {
```

File: src/utils/data/dataset.rs [hunk 28]
```diff
@@ -0,0 +1,277 @@
...
+        fn len(&self) -> u32 {
+            let mut label: Option<&OnMemData> = None;
+            if self.test {
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
            let mut label: Option<&OnMemData> = None;
            
            if self.test {
```

File: src/utils/data/dataset.rs [hunk 29]
```diff
@@ -0,0 +1,277 @@
...
+            write!(f, "{}{}", label_str, image_str)
+        }
+    }
+
+    /// MNIST dataset
+    ///
+    /// # Parameters
+    /// * `root` - Root directory of dataset
+    pub struct MNIST {
+        test: bool,
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
        is_test: bool,
```

File: src/utils/data/dataset.rs [hunk 30]
```diff
@@ -0,0 +1,277 @@
...
+            let mut label: Option<&OnMemData> = None;
+            if self.test {
+                if let Some(label_) = &self.test_label {
+                    label = Some(label_);
+                }
+            } else {
+                if let Some(label_) = &self.train_label {
+                    label = Some(label_);
+                }
+            }
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
                if let Some(label) = &self.test_label {
                    label = Some(label);
                }
            } else {
                if let Some(label) = &self.train_label {
                    label = Some(label);
                }
            }
```

File: src/utils/data/dataset.rs [hunk 31]
```diff
@@ -0,0 +1,277 @@
...
+            // big endian
+            // extract number of labels
+            if let Some(label_data) = label {
+                byte_arr_to_u32(&label_data.raw, 4, true)
+            } else {
+                0
+            }
+        }
+
+        fn get_item(&self, idx: u32) -> Result<MnistItem, Box<dyn error::Error>> {
```
Comment (utilForever) [inline src/utils/data/dataset.rs:220]: Where is the documentation for parameters?

File: src/utils/data/dataset.rs [hunk 32]
```diff
@@ -0,0 +1,277 @@
...
+                    image_data = Some(image_);
+                }
+            } else {
+                if let Some(label_) = &self.train_label {
+                    label_data = Some(label_);
+                }
+                if let Some(image_) = &self.train_img {
+                    image_data = Some(image_);
+                }
+            }
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
            if self.test {
                if let Some(label) = &self.test_label {
                    label_data = Some(label);
                }
                if let Some(image) = &self.test_img {
                    image_data = Some(image);
                }
            } else {
                if let Some(label) = &self.train_label {
                    label_data = Some(label);
                }
                if let Some(image) = &self.train_img {
                    image_data = Some(image);
                }
            }
```

File: src/utils/data/dataset.rs [hunk 33]
```diff
@@ -0,0 +1,277 @@
...
+
+            // build images vector
+            let mut image = Vec::new();
+            if let Some(image_) = image_data {
+                let size = n_rows * n_cols;
+                let start_idx = 16 + size * idx;
+                for idx in start_idx..start_idx + size {
+                    image.push(image_.raw[idx as usize]);
+                }
+            }
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
            let mut image_vec = Vec::new();
            
            if let Some(image) = image_data {
                let size = n_rows * n_cols;
                let start_idx = 16 + size * idx;
                
                for idx in start_idx..start_idx + size {
                    image_vec.push(image.raw[idx as usize]);
                }
            }
```

File: src/utils/data/dataset.rs [hunk 34]
```diff
@@ -0,0 +1,277 @@
...
+
+            // get label
+            let mut label = 0;
+            if let Some(label_) = label_data {
+                label = label_.raw[(8 + idx) as usize];
+            }
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
            let mut item_label = 0;
            
            if let Some(label) = label_data {
                item_label = label.raw[(8 + idx) as usize];
            }
```

File: src/utils/data/dataset.rs [hunk 35]
```diff
@@ -0,0 +1,277 @@
...
+
+            // build item
+            let item = MnistItem {
+                image,
+                label,
+                row: n_rows,
+            };
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
            let item = MnistItem {
                image: image_vec,
                label: item_label,
                row: n_rows,
            };
```

File: src/utils/data/dataset.rs [hunk 36]
```diff
@@ -0,0 +1,277 @@
...
+            Ok(item)
+        }
+    }
+
+    /// Test code for builtin datasets
```
Comment (utilForever) [inline src/utils/data/dataset.rs]: I think that it is unnecessary.

File: src/utils/data/dataset.rs [hunk 37]
```diff
@@ -0,0 +1,277 @@
...
+
+        #[test]
+        fn generate_mnist_dataset() {
+            if let Ok(mnist_test) = MNIST::new(PathBuf::from("raw"), true) {
+                let dataset_length = mnist_test.len();
+                assert_eq!(dataset_length, 10000);
+                if let Ok(dataset_item) = mnist_test.get_item(0) {
+                    println!("{:?}", dataset_item);
+                }
+            }
```
Comment (utilForever) [inline src/utils/data/dataset.rs]:
```suggestion
            if let Ok(mnist_test) = MNIST::new(PathBuf::from("raw"), true) {
                let dataset_length = mnist_test.len();
                assert_eq!(dataset_length, 10000);
                
                if let Ok(dataset_item) = mnist_test.get_item(0) {
                    println!("{:?}", dataset_item);
                }
            }
```

**Review by utilForever — CHANGES_REQUESTED** [review]
You should change many things. Please check them.
I'll review again after you resolved these issues.

File: .gitignore (hunk 1 above)
Comment (litcoderr) [reply .gitignore:13]: https://github.com/utilForever/polarlight/blob/cc85ee143a695b0622215c619f76617a06782b58/src/utils/data/dataset.rs#L118

`raw` directory is created inside the main project directory whenever the mnist test code is ran.

Should I add some comment in gitignore file

or 

should I change the test code to store mnist data somewhere else?

**Comment by litcoderr** [comment]
@utilForever I resolved all of your requested changes !!

**Review by utilForever — APPROVED** [review]
LGTM now. Thanks for you work! 🎉

**Pull request merged**